cdc/log.cc

```diff
@@ -416,6 +416,194 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
+                        updates.push_back({id, {}});
+                    }
+                    updates.back().cells.push_back({std::move(k), std::move(v)});
+                }
+
+                auto timestamp_and_ttl = std::make_pair(desc.tomb.timestamp, 0);
+                ups[timestamp_and_ttl].nonatomic_deletions.push_back({id});
+            });
+        });
+        return ups;
+    };
+
+    set_of_changes res;
+
+    auto& p = base_mutation.partition();
+
+    auto sr_updates = extract_row_updates(p.static_row().get(), column_kind::static_column);
+    for (auto& [k, up]: sr_updates) {
+        auto [timestamp, ttl] = k;
+        res[timestamp].static_updates.push_back({
+                ttl,
+                std::move(up.atomic_entries),
+                std::move(up.nonatomic_deletions),
+                std::move(up.nonatomic_updates)
+            });
+    }
+
+    for (const rows_entry& cr : p.clustered_rows()) {
+        auto cr_updates = extract_row_updates(cr.row().cells(), column_kind::regular_column);
+
+        const auto& marker = cr.row().marker();
+        auto marker_timestamp = marker.timestamp();
+        auto marker_ttl = marker.is_expiring() ? marker.ttl().count() : 0;
+        if (marker.is_live()) {
+            // make sure that an entry corresponding to the row marker's timestamp and ttl is in the map
+            (void)cr_updates[std::make_pair(marker_timestamp, marker_ttl)];
+        }
+
+        auto is_insert = [&] (api::timestamp_type timestamp, gc_clock::rep ttl) {
+            if (!marker.is_live()) {
+                return false;
+            }
+
+            return timestamp == marker_timestamp && ttl == marker_ttl;
+        };
+
+        for (auto& [k, up]: cr_updates) {
+            auto [timestamp, ttl] = k;
+
+            if (is_insert(timestamp, ttl)) {
+                res[timestamp].clustered_inserts.push_back({
+                        ttl,
+                        cr.key(),
+                        std::move(up.atomic_entries),
+                        std::move(up.nonatomic_deletions)
+                    });
+                if (!up.nonatomic_updates.empty()) {
+                    // nonatomic updates cannot be expressed with an INSERT.
+                    res[timestamp].clustered_updates.push_back({
+                            ttl,
+                            cr.key(),
+                            {},
+                            {},
+                            std::move(up.nonatomic_updates)
+                        });
+                }
+            } else {
+                res[timestamp].clustered_updates.push_back({
+                        ttl,
+                        cr.key(),
+                        std::move(up.atomic_entries),
+                        std::move(up.nonatomic_deletions),
+                        std::move(up.nonatomic_updates)
+                    });
+            }
+        }
+
+        auto row_tomb_timestamp = cr.row().deleted_at().regular().timestamp;
```

Comment:
You also need to consider the shadowable tombstone if cdc is ever to be enabled on a materialized view.

Reply:
Hint: tombs have an `operator bool()`, which returns true if `timestamp != missing_timestamp`.

Reply:
We won't use CDC on materialized views.
(and if we will, the code would probably need drastic changes anyway)

Reply:
Good point about the `operator bool()`